kitchen/src/kitchen/demos.clj

```diff
@@ -0,0 +1,144 @@
+  (printlog request "query params:" query-params)
+  (try
+    (let [{:strs [search]} query-params
+          search-url (str "https://en.wikipedia.org/w/api.php?action=query&prop=images&format=json&titles=" search)
+          image-titles (search->image-titles search)
+          _ (printlog request "image titles:" (vec image-titles))
+          image-entries (map image-title->image-entry image-titles)]
+      (printlog request "image entries:" (vec image-entries))
+      (utils/map->json-response {:result image-entries}))
+    (catch Exception e
+      (log/error e "encountered exception while performing image search")
+      (utils/exception->json-response e))))
+
+(defn- directory-location
+  [dir-name]
+  (let [base-dir (or (System/getenv "MESOS_DIRECTORY") (System/getProperty "user.dir"))]
+    (str base-dir File/separator dir-name)))
+
+(defn- copy-uri-to-file
+  [uri file]
+  (with-open [in (io/input-stream uri)
+              out (io/output-stream file)]
+    (io/copy in out)))
+
+(let [image-tagging-lock (Object.)]
+  (defn- retrieve-image-tags
+    [classifier-file image-location predictions]
+    (let [result-atom (atom nil)]
+      (locking image-tagging-lock
+        (reset! result-atom (sh/sh "python3"
+                                   classifier-file
+                                   (str "--image_file=" image-location)
+                                   (str "--num_top_predictions=" predictions))))
+      @result-atom)))
+
+(defn- image-tagging-handler
+  "Performs image tagging using tensorflow."
+  [{:keys [query-params] :as request}]
+  (printlog request "query params:" query-params)
+  (try
+    (let [{:strs [id name url predictions]} query-params
+          image-location (directory-location (str "images" File/separator name))
+          classifier-file-name "classify_image.py"
+          classifier-file (directory-location (str "images" File/separator classifier-file-name))]
+
+      (-> image-location str File. .getParentFile .mkdirs)
+      (-> classifier-file str File. .getParentFile .mkdirs)
+
+      (printlog request "downloading" url "to" image-location)
+      (copy-uri-to-file url image-location)
+
+      (printlog request "classifier resource:" (io/resource classifier-file-name))
+      (when-not (-> classifier-file str File. .exists)
+        (printlog request "saving" classifier-file-name "to" classifier-file)
+        (copy-uri-to-file (io/resource classifier-file-name) classifier-file))
+
+      (printlog request "classifying image" image-location)
+      (let [{:keys [err exit out]} (retrieve-image-tags classifier-file image-location predictions)
+            result-lines (filter #(str/includes? % "(score =") (str/split-lines out))]
+        (printlog request "classifier exit code:" exit)
+        (printlog request "classifier stdout:" out)
+        (printlog request "classifier stderr:" err)
+        (utils/map->json-response (cond-> {:id id, :out out, :predictions result-lines}
```

Comment:
Clean up files after processing is done.

Reply:
Now I delete the image files after processing.